src/waveform/renderers/waveformbeat.cpp

```diff
@@ -0,0 +1,72 @@
+#include "waveform/renderers/waveformbeat.h"
+
+#include "util/math.h"
+
+namespace {
+constexpr int kTriangleEdgeLength = 10;
+
+inline float getEquilateralTriangleAltitude(float triangleEdgeLength) {
+    return sqrt(3) / 2 * triangleEdgeLength;
+}
+} // namespace
+
+WaveformBeat::WaveformBeat()
+        : m_orientation(Qt::Horizontal), m_beatType(mixxx::track::io::BEAT), m_beatGridMode(1) {
+}
+
+void WaveformBeat::draw(QPainter* painter) const {
+    if (m_orientation == Qt::Horizontal) {
+        painter->drawLine(QPointF(m_position, 0), QPoint(m_position, m_length));
+        if (m_beatType == mixxx::track::io::BAR && m_beatGridMode == 1) {
+            // TODO: Get color from skin context
+            painter->setBrush(Qt::red);
+            painter->drawPolygon(getEquilateralTriangle(
+                    kTriangleEdgeLength, QPointF(m_position, 0), Direction::DOWN));
+            painter->drawPolygon(getEquilateralTriangle(
+                    kTriangleEdgeLength, QPointF(m_position, m_length), Direction::UP));
+        }
+    } else {
+        painter->drawLine(QPointF(0, m_position), QPoint(m_length, m_position));
+        if (m_beatType == mixxx::track::io::BAR && m_beatGridMode == 1) {
+            painter->setBrush(Qt::red);
+            painter->drawPolygon(getEquilateralTriangle(
+                    kTriangleEdgeLength, QPointF(0, m_position), Direction::RIGHT));
+            painter->drawPolygon(getEquilateralTriangle(kTriangleEdgeLength,
+                    QPointF(m_length, m_position),
+                    Direction::LEFT));
+        }
+    }
+}
+
+QPolygonF WaveformBeat::getEquilateralTriangle(float edgeLength,
```

Comment:
It looks like this doesn't depend on any class members, so it can be moved to the anonymous namespace instead of being a private class member.